In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

data= pd.read_csv("/content/Default_Fin.csv")
data.head()


Index  Employed  Bank Balance  Annual Salary  Defaulted?
0      1         1       8754.36      532339.56           0
1      2         0       9806.16      145273.56           0
2      3         1      12882.60      381205.68           0
3      4         1       6351.00      428453.88           0
4      5         1       9427.92      461562.00           0

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Index          10000 non-null  int64  
 1   Employed       10000 non-null  int64  
 2   Bank Balance   10000 non-null  float64
 3   Annual Salary  10000 non-null  float64
 4   Defaulted?     10000 non-null  int64  
dtypes: float64(2), int64(3)
memory usage: 390.8 KB


In [ ]:
data.describe()

Index      Employed  Bank Balance  Annual Salary    Defaulted?
count  10000.00000  10000.000000  10000.000000   10000.000000  10000.000000
mean    5000.50000      0.705600  10024.498524  402203.782224      0.033300
std     2886.89568      0.455795   5804.579486  160039.674988      0.179428
min        1.00000      0.000000      0.000000    9263.640000      0.000000
25%     2500.75000      0.000000   5780.790000  256085.520000      0.000000
50%     5000.50000      1.000000   9883.620000  414631.740000      0.000000
75%     7500.25000      1.000000  13995.660000  525692.760000      0.000000
max    10000.00000      1.000000  31851.840000  882650.760000      1.000000

Missing Values


In [ ]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy="mean")

data["Bank Balance"] = imputer.fit_transform(data[["Bank Balance"]])
data["Annual Salary"] = imputer.fit_transform(data[["Annual Salary"]])

data.describe()


Index      Employed  Bank Balance  Annual Salary    Defaulted?
count  10000.00000  10000.000000  10000.000000   10000.000000  10000.000000
mean    5000.50000      0.705600  10024.498524  402203.782224      0.033300
std     2886.89568      0.455795   5804.579486  160039.674988      0.179428
min        1.00000      0.000000      0.000000    9263.640000      0.000000
25%     2500.75000      0.000000   5780.790000  256085.520000      0.000000
50%     5000.50000      1.000000   9883.620000  414631.740000      0.000000
75%     7500.25000      1.000000  13995.660000  525692.760000      0.000000
max    10000.00000      1.000000  31851.840000  882650.760000      1.000000

In [ ]:
def winsorize_column(column, lower_limit, upper_limit):
    column_copy = column.copy()
    p = np.percentile(column_copy, [lower_limit, upper_limit])
    column_copy[column_copy < p[0]] = p[0]
    column_copy[column_copy > p[1]] = p[1]
    return column_copy

data["Bank Balance"] = winsorize_column(data["Bank Balance"], 5, 95)
data["Annual Salary"] = winsorize_column(data["Annual Salary"], 5, 95)

Normalización

In [ ]:
from numpy.random.mtrand import standard_cauchy
from sklearn.preprocessing import StandardScaler

X = data[["Bank Balance","Annual Salary"]]
y = data["Defaulted?"]

scaler = StandardScaler()
X_scaled = scaler.fit(X)


Modelo

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

logistic= LogisticRegression(solver="lbfgs")
logistic.fit(X_train,y_train)
y_pred = logistic.predict(X_test)


In [ ]:
result = classification_report(y_test,y_pred,zero_division=1)
print("Classification Report")
print(result)

Classification Report
              precision    recall  f1-score   support

           0       0.97      1.00      0.98      1931
           1       1.00      0.00      0.00        69

    accuracy                           0.97      2000
   macro avg       0.98      0.50      0.49      2000
weighted avg       0.97      0.97      0.95      2000



* Precision:
Para 0, los datos que se supone que deberían ser cero, los clasifica así el 97% de las veces, para 1 el 100%.
* Recall:
Para la clase 0, el recall es 1.00, lo que indica que el 100% de las muestras de la clase 0 fueron correctamente clasificadas como tal.
 Sin embargo, para la clase 1, el recall es 0.00, lo que indica que ninguna muestra de la clase 1 fue correctamente clasificada como tal. Este resultado sugiere que el modelo no pudo identificar correctamente ninguna muestra de la clase 1.
* F1 score:
Para la clase 0, el F1-score es 0.98, lo cual es alto y muestra un buen equilibrio entre precisión y recall. Sin embargo, para la clase 1, el F1-score es 0.00, lo que indica que el modelo no logró clasificar correctamente ninguna muestra de la clase 1.

* promedio ponderado:
Para el promedio ponderado, la precisión, recall y F1-score son 0.97, lo que indica un buen rendimiento general del modelo en todo el conjunto de datos.



In [ ]:
df = pd.read_csv("/content/healthcare-dataset-stroke-data.csv")
df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

In [ ]:
df.isnull()

id  gender    age  hypertension  heart_disease  ever_married  \
0     False   False  False         False          False         False   
1     False   False  False         False          False         False   
2     False   False  False         False          False         False   
3     False   False  False         False          False         False   
4     False   False  False         False          False         False   
...     ...     ...    ...           ...            ...           ...   
5105  False   False  False         False          False         False   
5106  False   False  False         False          False         False   
5107  False   False  False         False          False         False   
5108  False   False  False         False          False         False   
5109  False   False  False         False          False         False   

      work_type  Residence_type  avg_glucose_level    bmi  smoking_status  \
0         False           False              False  False           False   
1         False           False              False   True           False   
2         False           False              False  False           False   
3         False           False              False  False           False   
4         False           False              False  False           False   
...         ...             ...                ...    ...             ...   
5105      False           False              False   True           False   
5106      False           False              False  False           False   
5107      False           False              False  False           False   
5108      False           False              False  False           False   
5109      False           False              False  False           False   

      stroke  
0      False  
1      False  
2      False  
3      False  
4      False  
...      ...  
5105   False  
5106   False  
5107   False  
5108   False  
5109   False  

[5110 rows x 12 columns]

Se encuentran solo valores nulos en la columna BMI

In [ ]:
valores_nulos = df.isnull().sum()
print(valores_nulos)

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64


In [ ]:
df = df.dropna()

In [ ]:
valores_nulos = df.isnull().sum()
print(valores_nulos)

id                   0
gender               0
age                  0
hypertension         0
heart_disease        0
ever_married         0
work_type            0
Residence_type       0
avg_glucose_level    0
bmi                  0
smoking_status       0
stroke               0
dtype: int64


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4909 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 4909 non-null   int64  
 1   gender             4909 non-null   object 
 2   age                4909 non-null   float64
 3   hypertension       4909 non-null   int64  
 4   heart_disease      4909 non-null   int64  
 5   ever_married       4909 non-null   object 
 6   work_type          4909 non-null   object 
 7   Residence_type     4909 non-null   object 
 8   avg_glucose_level  4909 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     4909 non-null   object 
 11  stroke             4909 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 498.6+ KB


In [ ]:
from sklearn.svm import SVC
from sklearn.preprocessing import LabelEncoder

x = df[["gender","age","hypertension","heart_disease",
        "ever_married","work_type","Residence_type",
        "avg_glucose_level","bmi","smoking_status"]]

y = df["stroke"]#Target

le = LabelEncoder()

x["gender"] = le.fit_transform(x["gender"])
x["ever_married"] = le.fit_transform(x["ever_married"])
x["work_type"] = le.fit_transform(x["work_type"])
x["Residence_type"] = le.fit_transform(x["Residence_type"])
x["smoking_status"] = le.fit_transform(x["smoking_status"])

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

model = SVC()
model.fit(x_train,y_train)
y_pred=model.predict(x_test)



In [ ]:
result_svc = classification_report(y_test,y_pred,zero_division=1)

print(result_svc)

              precision    recall  f1-score   support

           0       0.95      1.00      0.97       929
           1       1.00      0.00      0.00        53

    accuracy                           0.95       982
   macro avg       0.97      0.50      0.49       982
weighted avg       0.95      0.95      0.92       982



In [ ]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, y_pred)
print(accuracy)

0.9460285132382892


En general, la precisión promedio del modelo es del 95%, lo que indica que la mayoría de las predicciones son correctas. Sin embargo, dado que el recall promedio es bajo (0.50) y la puntuación F1 promedio es baja (0.49), se puede decir que el modelo tiene dificultades para identificar la clase 1 y tiene un rendimiento deficiente en términos de clasificación general.